### DISCUSSION TITLE
Help with USB Power Trigger

### DISCUSSION THREAD

**Discussion by nerdfever** [gopro/labs#912] [General]
I'm new with this stuff; please be gentle.

GoPro 12 Hero, today's Labs firmware (2024-05-15). 1 TB SanDisk Ultra card (seems to work OK).

I made a QR code "mSr5Xp30fW0e1!u5N!S!u5E!R".

Intent is Standard video, 5K 8:7, 30 fps, Wide, Hypersmooth ON, High bitrate, 10 bit, start on USB power after 5 seconds, stop on USB power removed after 5 seconds.

As soon as I show the QR code to the camera it shuts down. I then unplug the USB, wait a few seconds, and plug it back in.

**Nothing happens.** 

I expected it to start recording in that mode after 5 seconds, then stop 5 seconds after USB power goes away. 

And again each time USB power comes on/goes off.

I've tried it a few different ways - (just the USB power commands, different delays, etc.). Same thing - camera shuts down when I show it the QR code, then won't start up again unless I press the buttons on it. Regardless of plugging/unplugging USB power.

_What am I doing wrong?_

![GoProQR: mSr5Xp30fW0e1!u5N!S!u5E!R Standard Wide](https://github.com/gopro/labs/assets/9833299/15735da8-3cf7-425c-b05a-8b91054589fc)

**Comment by nerdfever** [comment]
GitHub isn't letting me edit my post... When I show the QR code to the camera it immediately says "Shutdown until USB power" and shuts down. Same thing happens regardless of whether the camera is on USB power or battery at the time I show it the QR. 

If remove/replace USB power (or plug it in if I'd been on battery)...nothing happens. Just the red charging light comes on.

The camera seems to work OK otherwise, so I suppose it's something I'm misunderstanding...but what?

**Reply by dnewman-gpsw** [reply]
The !uS style codes are very old and haven't been tested on HERO12.  The reason is the newer methods are better, although more complex, as they are more powerful. It has been written up on this forum before. Best to use the Blockly interface to learn this stuff https://gopro.github.io/labs/build/ 

The basics:
*WAKE=2  <- set up the camera to auto boot of power

Then setup a boot command for what action the camera does on boot.
*BOOT="!Lbt"!SAVEbt=your commands here

Then USB power conditionals work:
>u0!S<u0!5E - if USB power Start     if USB not powered end after 5 seconds.

**Comment by nerdfever** [comment]
Thanks; I'll look into that. I just Googled on "GoPro Labs" and "USB Power" and found that stuff. 

In particular I found https://gopro.github.io/labs/control/usb/ 

I had no idea the system I was using was deprecated. GoPro should really update their (many) pages discussing the old system with notes about that.

**Reply by dnewman-gpsw** [reply]
Users can update, many do offer contributions to the documentation. This is just a wiki of sorts, Labs is intentionally light of documentation, as it if we made it polished, that would take more time, and you would get fewer features. Also this feature is not rarely deprecated, as it will work on older cameras, Labs is widely user from HERO7 onward. HERO11/12 just have better solutions, but support have the support old command (so a bug is likely.)

**Reply by nerdfever** [reply]
The whole business with "BOOT" and "L" and "WAKE" and !M and oM not working is really confusing - the existing pages like https://gopro.github.io/labs/control/extensions/ give no hint that the instructions there are deprecated and don't work on the lastest camera (HERO12). I've been struggling with this for 2 days and still can't make anything work on the HERO12. At least if a page has obsolete instructions take it down or put a note in it. Please! (I"m not alone, see: https://www.reddit.com/r/goprohack/comments/1c75bhg/how_to_start_hero12_black_when_power_to_usb/ )

... added ...

That reddit post implies that files stored on the microSD card (it mentions files "qr.bin" and "bt" ) are somehow releavant, but none of the pages here on github say anything about where anything is actually stored - they just talk about QR codes. Where does the camera store the info from the QR codes? Is it in files on the microSD card? Which ones? How long does that info stay valid? Or is it all stored in RAM or NVRAM inside the camera?

Lots of details are given but the basic concepts never are, and having multiple obsolete commands described that don't work on the current hardware doesn't help.

Also some documentation seems to be self-contradictory. In one place it says !R "repeats the whole command" (which command is that? Not clear) but in another place it implies !R does a Reboot (the example is !RESET!10R, with no explanation of what the 10 part does.)

I get the feeling this all made sense and was pretty simple originally, but then things changed and nobody updated the docs to match. It makes perfect sense to someone who already knows how the pieces fit together, but there's no explicit explanation for newcomers.

Reading between the lines I gather the first character of commands (!, *, $, o, ^ and maybe others) means something, but nowhere is anything clearly stated in one place. I'm trying to write up my own documentation as I read (what I have so far is attached FYI), but I'm so far unable to put it together to get even a simple dashcam working.

</rant> (sorry)

[My command notes.txt](https://github.com/gopro/labs/files/15358827/My.command.notes.txt)

...  more added ...

As recently as April 12 2024, you posted "Try adding !MTUSB=1 this will bypass the power support current testing." Yet I think the !M syntax is deprecated and the correct syntax is now *M. I think. It's really hard to follow instructions that use syntax that's obsolete...could you please clarify?

**Reply by nerdfever** [reply]
Let me ask this another way. Will this work on the latest labs firmware on a HERO 12? This is the best I've been able to figure out:

*WAKE=2!TUSB=1*BOOT="!Lboot"!SAVEboot=u0mSr5Xr30fW0e1b1d10!S<u0!5E!O 

Intent is to wake on USB, ignore USB (just plain 5 volts on the USB interface), then set mS r5X r30 fW0 e1 b1 d10 (Standard video, 5k 8:7, 30 fps, Wide lens, Hypersmooth ON, high bitrate, 10 bits/pixel) and start recording.

When USB power goes away, end recording and shutdown.

If USB power comes back later, I'd like it to start another recording. No time limit (it's a 1 TByte microSD card and on USB power so it should be able to go for days).

Ideally I'd like it to work without any battery at all (less heat and risk of swelling) but I can leave the battery in if I have to.

I tried it, it doesn't seem to work. I get messages like "Feature activated WAKE=2" (even before I showed it any QR code!!) and LBoot. I didn't ask for any messages.

My log looks like this:

[2024-05-17 20:59:19 82%   ] Labs:Main Start 
[2024-05-17 20:59:19 82%   ] Labs:H23.01.02.20.70 
[2024-05-17 20:59:19 82%   ] Labs:Initialize 
[2024-05-17 20:59:19 82%   ] Read QR:*WAKE=2!TUSB=1*BOOT="!Lboot"!SAVEboot=u0mSr5Xr30fW0e1b1d10!S<u0!5E!O 
[2024-05-17 20:59:21 82%   ] STARTCAPTURE:0 
[2024-05-17 20:59:22 82%   ] Save Script:C:\MISC\boot 
[2024-05-17 20:59:35 82%   ] capture stopped:user input or error 
[2024-01-01 00:00:00-delayed ] wake reason:USB1_INSERTED 
[2024-05-17 20:59:53 ->--% ] Labs:Display Messages 
[2024-05-17 20:59:57 ->82% ] Features activated:

**Reply by nerdfever** [reply]
Sorry - forum software seems to eat some of the commands.

Here's what I tried: `*WAKE=2!TUSB=1*BOOT="!Lboot"!SAVEboot=u0mSr5Xr30fW0e1b1d10!S<u0!5E!O`

**Reply by dnewman-gpsw** [reply]
You need to start with a simpler without boot commands. As you haven't grokked the command perfectly, you are making errors, and putting errors in boot commands is not great.  All commands are either actions or camera settings.  Actions always start with ! , settings are lower case followed by parameters. This is in the docs.  So !S is an action to start, r4 is a setting of resolution 4k.  This has been the case from day one.  The command language has evolved with conditionals, branches and 4CC memory stores. This is what is giving you trouble.  Old command was !MTUSB=1 for permanent memory store in TUSB the value 1.  Also oMTUSB=1 is the temporary version, it is not obsolete, works on all cameras. The problem was the overuse of ! , so in 2024 firmware memory got the new * and $ storage types. That is it.  

So !TUSB is not a command of you follow the above.

**Comment by dnewman-gpsw** [comment]
Your command core will not work the way you want 

!S<u0!5E!O

This means:
Start();
If(no USB power) end(5);
Shutdown();

So it will start and immediately shutdown or after 5s.

**Reply by nerdfever** [reply]
ok, thanks. What _should_ I be doing then? What's the correct command string?

**Comment by dnewman-gpsw** [comment]
Use https://gopro.github.io/labs/build/ as it knows the coding rules.

Corrections for some of the errors above, as these might lend others down the path:

!R is repeat, !O is Off, !OR is Off then Restart, not repeat. You have a typo, !1OR not !10R. These are not contradictions. All in the docs https://gopro.github.io/labs/control/actions/
 
!SAVEfilename=data is the only user controllable function to store things to the SD card (in MISC). Other files are automatically created as needed, for internal purposes, you do not need to touch them. 

It is documented that !Mwxyz= or *wxyx= are permanent saves to NVM, and oMwxyz= or $wxyz= saves to RAM. All within the above actions docs.

The command language is designed for compactness, not ease of use, as QR code as limited in capacity.  It is fairly logical; re-read docs from the top here https://gopro.github.io/labs/control/tech

**Reply by nerdfever** [reply]
I have this much working: `*WAKE=2*TUSB=1`  That makes it wake up on USB power and trust the USB power. After that I'm able to set the mode but not start recording. I tried this and it sets the mode but recording doesn't start:

`*WAKE=2*TUSB=1mSr5Xr30fW0e1b1d10!S`

I also tried this, doesn't work either:

`*WAKE=2*TUSB=1*BOOT="!Lboot"!SAVEboot=u0mSr5Xr30fW0e1b1d10!S<u0!5E!O`

I don't have a battery in the camera, so what it does after USB power goes away doesn't matter (it has to shut off).

BTW, what does "!R" repeat then? The last command? All commands? Starting from which one?

...added...

The last command above almost works. It starts on USB power, sets the mode and starts recording. But it stops after recording for 2 seconds and powers off. It doesn't say why (no complaint about not enough USB power). The USB power supply has plenty of power (a 3 amp bench supply).

**Reply by dnewman-gpsw** [reply]
Some of these question are so easy to answer through experimentation.  Start simple, test each command one at a time. You can do it. You will find it has typos. Like what is ```r30``` (30K we don't have that resolution yet.) ```p``` is for f**p**s. 
![image](https://github.com/gopro/labs/assets/10856084/d1b0d2ef-64fb-4a2f-b40a-ad3aa1d6e249)

This works every time on my HERO12.

**Reply by dnewman-gpsw** [reply]
Sorry for the p25, I'm traveling in Europe at the moment.

**Reply by dnewman-gpsw** [reply]
> But it stops after recording for 2 seconds and powers off. It doesn't say why (no complaint about not enough USB power). The USB power supply has plenty of power (a 3 amp bench supply).

I explained that earlier, as that is what you ask the camera to do, shutdown immediately after starting capture, nothing to do with power.  <u0 is not a wait command, is condition on the power status now, it will continue immediately. Practice with print statements

**Comment by nerdfever** [comment]
Hey I think I got it working!!

`*WAKE=2*TUSB=1*BOOT="!Lboot"!SAVEboot=mSr5Xr30fW0e1b1d10!S`

That's working on a Hero 12 without a battery, with the latest firmware (as of May 2024).

I took out all the stuff about conditional on USB power, since I don't even have a battery in there.

David, I really appreciate the replies, but the existing documentation is profoundly confusing - I interpreted the whole QR code as getting stored and executed on boot. That's not what happens (I think).

Here's a summary for others of how this works (I think). This is the "missing piece" I needed to understand (I'm writing this for myself and for others who may be confused):

The QR is executed once only when the camera sees it. 

If you want anything to happen later on boot, you must use `*BOOT` in the QR code to tell the camera to load and run a script at boot time (I called the script "boot", but you can call the script whatever you want).

The * commands don't need to be in the script to run later - the camera will remember those. But everything else does need to be in the script.

My interpretation of what's going on is that `*WAKE=2 `tells it to wake on USB power, and the * means remember that across power cycles.

So when the camera sees the QR code `*WAKE=2*TUSB=1*BOOT="!Lboot"!SAVEboot=mSr5Xr30fW0e1b1d10!S` that does exactly 4 and only 4 things (even tho there's like a dozen commands in there):

1 - The camera remembers `WAKE=2` (wake on power)

2 - The camera remembers `TUSB=1` (trust USB)

3 - The camera remembers `BOOT="!Lboot"`, which means when the camera boots it's supposed to load and run the "boot" script.

4 - Last, the camera does `!SAVEboot=mSr5Xr30fW0e1b1d10!S`, which means it saves a script called "boot" with the contents "mSr5Xr30fW0e1b1d10!S".  (I think it's saved in a file on the microSD card...haven't verified that yet.)

That's it - those 4 things. The QR code in the example doesn't set any modes or start recording.

But, later when the camera boots, THEN it loads and runs then "boot" script ("mSr5Xr30fW0e1b1d10!S"), which sets the modes and starts recording. THEN. Not when the camera saw the QR code.

If the documentation said all this somewhere I'd have figured it out days ago. I think if you post this message in a sticky, it'll help a lot of people.

Regards...

**Reply by nerdfever** [reply]
Also, my revised notes are attached.


[GoPro Labs command notes for HERO12.txt](https://github.com/gopro/labs/files/15366656/GoPro.Labs.command.notes.for.HERO12.txt)

**Comment by nerdfever** [comment]
BTW, the documentation is a little ambiguous about whether you need double quotes around fourcc assignments. In one place it says you do, but other examples (esp. for BOOT) don't use them, and it seems to work.

Also, to reduce confusion, "BOOT" is not a "command". BOOT is a metadata variable. Only things prefixed with ! are "commands".

**Reply by dnewman-gpsw** [reply]
The quote rules for metadata are documented, quotes for strings and without for numbers. Only !SAVE, no M or * as it is not a metadata command, it doesn't use quotes as it can store them.  

Yes BOOT is a metadata variable with extra power.  It's how they all work, BITR, EXPT, NR01, ... there is about 100.  But you can also do *MINE="My camera".  While it has no extra power, all are stored as metadata in every file. That is why they metadata commands. Using only upper case and numbers, the FourCC supports over a million permutations.

**Reply by dnewman-gpsw** [reply]
Why close, as you still haven't completed the coding loop for the USB power test?

**Comment by nerdfever** [comment]
Here's a cleaner version. 

![Dashcam QR blockly](https://github.com/gopro/labs/assets/9833299/847b64d3-d6b6-4aae-8b0c-1548c03e5372)



It works with the battery in the camera, so it can properly close the video with !E when USB power goes away. 

It also waits 30 seconds after USB power before starting capture, to charge the battery a little each time.

If stared via button press on battery power, it skips the auto capture and the GoPro can be used for other things.

I don't know if the 1 second pause in the loop (and after ending the capture) is necessary, but I saw other people using them and put them in. They seem harmless enough.

Last, note that the blockly code generates the QR:

```!MBOOT="!Lbt"!SAVEbt=<u0"On battery"+!X"Charging..."!30NmSr5Xr30fW0e1b1d10!S!1N<u0!E+!1N+!1O!R55```

To make it work on the HERO12 with current firmware I changed the start to use the * syntax and added TUSB (which I didn't see as an option in the blockly page):

`*WAKE=2*TUSB=1*BOOT="!Lboot"!SAVEboot=<u0"On battery"+!X"Charging..."!30NmSr5Xr30fW0e1b1d10!S!1N<u0!E+!1N+!1O!R55`

BTW, the scripting language is _definitely_ Turing complete.

**Reply by dnewman-gpsw** [reply]
!M is still used in H12, so you don't need to convert. Good news you got this working.

**Comment by ghost** [comment]
I've been able to get the actions AFTER boot to work, but I have to boot it manually. How do I get it to boot on USB power and where do I add in those commands since they don't seem to be in blocky?

**Reply by dnewman-gpsw** [reply]
Try Blockly does yet list the extensions, but under settings you can use "Add QR Command". For boot on power use *WAKE=2 (or just do that once, as you shouldn't need it in the Blockly script.)

**Comment by ghost** [comment]
Forgive me, but I'm new at this and special thanks to @nerdfever for helping me visualize it...

Like this?

<img width="453" alt="image" src="https://github.com/gopro/labs/assets/172081338/e14d7c9a-8021-4fe7-bd93-cf930554b269">

**Reply by nerdfever** [reply]
That's pretty much it, except the *WAKE=2 isn't part of the script. That gets executed once and the camera remembers it. The rest of it goes into the script you want to run on boot (I called mine "boot").

Don't get the QR code confused with the script (they are easy to confuse). The QR code is shown to the camera just once and it needs to do 2 things: First, setup the camera to wake on USB power and run a script when it does. Second, save a script (the one that will be run on boot). Those are 2 different things that need to be in the QR.

I'm still fiddling with some aspects of this myself - I'll post here when I figure out more.

Here (attached) is what I've learned so far.
[GoPro Labs command notes for HERO12.txt](https://github.com/user-attachments/files/15745462/GoPro.Labs.command.notes.for.HERO12.txt)

**Reply by ghost** [reply]
Thanks...tested and working!

I am a rank amateur here so I appreciate all work being done that I can learn from

**Reply by ghost** [reply]
This was working last night, but when I went out to the car and started it up this morning, it booted up, but didn't run the post boot.

Any suggestions? I note that yours runs !Lboot versus !Lbt. Does that make a difference?

**Reply by dnewman-gpsw** [reply]
No difference.  The boot script is on the SD card, so if you format or change the SD the script will not be there to run. Is the only way I know that a script will not run.

**Reply by nerdfever** [reply]
Re `!Lboot` versus `!Lbt`, what comes after !L is the name of the script, so it's "boot" vs "bt". The name doesn't matter but the one you load with `!L` and the one you save with` !SAVE` need to be identical. You're saving the script on the SD card and then loading it later at boot. Post the whole QR code you're using and perhaps I can give some advice.

**Comment by ghost** [comment]
Thanks for your help. I really appreciate it...

*WAKE=2!MBOOT="!Lbt"!SAVEbt=<u0!X!S<u0!E+!1O!R7

**Comment by ghost** [comment]
After reading your notes, I added *TUSB=1 just in case the USB provider power was the issue

`*WAKE=2*TUSB=1!MBOOT="!Lbt"!SAVEbt=<u0!X!S<u0!E+!1O!R7`

**Reply by ghost** [reply]
UPDATE: This has worked 2 times using USB charger in house and 3 times using USB charging in car. We'll see how it goes from here

**Reply by ghost** [reply]
Hold on...I think I know what happened. After I got done configuring it and getting it working, I used the "Delete All Files from SD Card" to delete all the junk videos I made, which blew away my script file (DUH!). Lesson learned

**Comment by nerdfever** [comment]
Hi David - a few clarifying questions:

1 - What is the correct command to set lens? The documentation says fW fL or fH (wide, linear, or linear+horizontal level, respectively), but the QR code generator generates fW0 fL0 and fH0. Should the trailing zeros be there or not? (Perhaps it works either way.)

2 - What's the difference between command !nN (pause n seconds) and !nNQ (pause "approximately" n seconds)? When would one be preferred over the other?

3 - Exactly where can I find GoPro's documentation re which modes may validly be combined (on the Hero 12)? For example the QR code generator lets me combine mode r5X (5k 8:7 full sensor mode) with fL (linear lens), but I think the camera in fact can't do both at the same time - I think r5X works only with fW (Wide lens). Where is it documented which modes can be combined?

Last, for those who are trying to construct their own QR codes, the !R command (repeat) appears to in fact be a "goto" command of the form !Rggg where ggg (if present) refers to the character offset into the script where script execution should resume. (If there is no ggg, it jumps to offset 0, the start of the script). So !R22 means jump to (goto) character 22 into the script (where position 0 is the first character).
I don't think this is written down anywhere - I figured it out by reverse engineering the blockly thing.

**Reply by dnewman-gpsw** [reply]
1 - the 0 is option, it is the amount of zoom.
2 - For short <10seconds they make no difference.  When you compare !60N and !60NQ  -  The first wake a few seconds before and then pauses for precise-r timing, wheras NQ version execute immediate, with less concern about precision.  NQ was for power optimization.
3 - For that you need to read your camera manual.  Labs is following what the camera can do.

Yes !Rxx is fairly new, added for Blockly support.

**Comment by nerdfever** [comment]
I think this will be my final post on this thread.

After some experimentation, I discovered:

* The Hero 12 DOES execute the saved script once when it sees a new QR (even tho it's in a SAVE= command)
* There appears to be no way to stop script execution externally (button, etc.)
* The Rnn command jumps to offset nn in current script (0 is first position)
* If you get stuck in a loop where the camera won't read a new QR,
  remove the SD card and delete \MISC\bt (or any script you saved) and \MISC\qr.bin
* Conditionals execute ONE command. Use + to concatenate multiple commands into one.

Also the !N and !NQ commands are practically limited to 9 seconds; any longer and the camera shuts down.
But you can use as many as you like at 9 seconds each (I use 5 to pause for 45 seconds).

* Pause with !10N or !10NQ (any value >= 10 seconds) shuts down camera
* Pause with up to !9 works. OK to use multiple !9N to get longer times.

Here's the QR code I'm using. It's tested and works:

`*WAKE=2*TUSB=1*BOOT="!Lbt"!SAVEbt=<u0"On battery"+!XmSr5Xp30fWe1b1d10"Standard video\nCharging 45s..."!9N!9N!9N!9N!9N!S!3N<u0"Shutting down"+!E+!2O<r0"Exiting script"+!1N+!X!R85`
                               
Broken down:

`*WAKE=2 `			Remember in flash; Wake on power (any)
`*TUSB=1`				Remember in flash: Trust USB power (do not reject USB power based on capabilities)
`*BOOT="!Lbt`"		Remember in flash: Load and run script "bt" on boot
`!SAVEbt=`			Save following to script named "bt"
`<u0"On battery"+!X`	If NOT on USB power, print "On battery", then exit script 
`mS	`				Mode Standard Video
`r5X	`				Mode 5k Extreme 8:7
`p30	`				30 frames/second
`fW		`			Wide lens
`e1`					Hypersmooth ON
`b1`					High bitrate (~ 120 Mbps)
`d10`					10 bits/pixel (HEVC, H.265)
`"Standard video\nCharging 45s..."`
					Print text.
`!9N!9N!9N!9N!9N`		Wait 45 seconds (9 seconds five times - this lets the battery charge 45 seconds each time; it does not charge while recording)
`!S`					Start recording
`!3N`					Wait 3 seconds
`<u0"Shutting down"+!E+!2O`
					If not on USB power, print "Shutting down", close video, wait 2 seconds, turn camera off
`<r0"Exiting script"+!1N+!X`
					If not recording (can happen only if user pressed record button on top), print "Exiting script", wait 1 second, exit script.
`!R85	`			Jump to 85 characters from start of script (this is the !3N above).

Attached are my final
[GoPro Labs command notes for HERO12.txt](https://github.com/user-attachments/files/15794151/GoPro.Labs.command.notes.for.HERO12.txt)
 notes which supplement that GoPro documentation. It includes Python code that produce the bar codes as PDFs.

**Reply by dnewman-gpsw** [reply]
One small correction 
 > The Hero 12 DOES execute the saved script once when it sees a new QR (even tho it's in a SAVE= command)

This is not the case. If you do !SAVEbt="Hello World", nothing happens.  However the *BOOT="!Lbt"!SAVEbt="Hello World" is executed.  I think BOOT should be fixed to only execute on boot, which will clear up the confusion.

**Reply by nerdfever** [reply]
"fixed" by whom? Me? GoPro? Somebody else?

**Reply by dnewman-gpsw** [reply]
Me, at GoPro.

**Reply by nerdfever** [reply]
Yes, executing only on the load command would be better. I just realized you're the guy responsible for the labs QR code stuff - thank you for doing it, it's really useful. The documentation online could be much clearer, but the code itself works as it should.

**Reply by dnewman-gpsw** [reply]
Please contribute the documentation wherever you feel it is missing information.  All the pages are editable, and I can approve and merge anyone's additions.

**Comment by nerdfever** [comment]
For a start, put it all in one page. Having lots of scattered pages makes things difficult to find.

**Reply by dnewman-gpsw** [reply]
... there are many dozens of pages, that is a lot to parse in one reading. It is was split up to be less overwhelming.

**Reply by nerdfever** [reply]
Many dozens == easy to miss one (or several). If they're all in one place you know you've got it all. Organization is the key.

Also, one issue I had was the several times I'd find stuff in one of the pages and then post about it here, and you'd tell me ="oh, it doesn't work that way any more since [revision of <date>]". But the pages still have the old info.

It all took a lot of figuring out that would have been unnecessary if (a) all the info were in one place, and (b) it was edited to be up to date.

**Reply by dnewman-gpsw** [reply]
"If they're all in one place.." - then make a single page, give it a go.  I wouldn't call a single page of Labs document organized, but prove me wrong.

"But the pages still have the old info" and you are welcome to add those notes. Also the information is rarely old, rather it applies to only some camera models. So removing it would upset the GP1 camera users, for which there are many.

The problem is Labs is an engineering only effort, 90% of the work is focused on the engineering new abilities, not marketing or documentation. More time spent on web page and documentation is less time spent in creating new stuff.  So I'm looking for more external collaborators, half a dozen already contribute.

**Reply by dnewman-gpsw** [reply]
BTW - it is all in one place https://gopro.github.io/labs/ and somewhat organized.

**Reply by nerdfever** [reply]
I definitely understand the problem re limited resources. When I look at that page I see this part:
![image](https://github.com/gopro/labs/assets/9833299/1cd961be-c2c5-44e5-8ded-4647f159a67a)

It's unfortunate that the headings are the links are in the same shade of blue - it makes it difficult to tell exactly how many links are there.

But I think the answer is 5: "Camera macros", "Camera Extensions", "Visual Script Tool", "Action Commands", and "Settings Commands".

Is that correct? If not, where are ALL the pages listed and indexed?

Or do you mean each of the links in the big table below that (that's a lot of pages!)? FWIW, it's also easier to keep things updated if everything is in one page, so the maintainer can easily do ^F to find all occurrences of a reference that needs changing, instead of looking in dozens of pages.

Why not just make the single page at https://gopro.github.io/labs/ contain everything, with the first part (that's already there) as an index to what's below?

**Reply by dnewman-gpsw** [reply]
I just don't agree that would be easier.  All those 6 links (I see 6 in the your screen grab) have links within them, all somewhat organized. I see two shades of blue, but yes they are similar, that was CCS provided by GoPro. I'm sure that too is editable. The links do underline when you mouse over them, so no hiding that they are links. There is not single common usage for Labs, so really isn't any reason anybody should read everything.

**Reply by dnewman-gpsw** [reply]
Again the documentation is all open source, and this repo has been forked 8-9 times by others, so you can be the 10th. You can make your example Labs single page in the style you like, all free on github.

**Reply by Lin-Admin** [reply]
>  this repo has been forked 8-9 times by others

I was not able to find forks, could you please give me a few URLs?

**Reply by dnewman-gpsw** [reply]
In the standard place for github : https://github.com/gopro/labs/forks No idea what these forks offer.  I don't manage them.

**Reply by remltr12** [reply]
I am hoping to start on USB power without a battery. Will that work or not? What I ultimately want is to boot the camera on usb power and immediately record video 2.7K with super view and stop when usb power is lost. Rinse and repeat. Any help would be greatly appreiciated. I am not looking to learn a new coding language btw.

**Reply by dnewman-gpsw** [reply]
That is straight forward for most GoPro Labs cameras.  You do have to mention what camera.  Also the the downside of no battery, is the files will not be closed, and will need to be repaired on the next boot.

**Reply by remltr12** [reply]
I am using a GoPro 11 Black.

**Reply by nerdfever** [reply]
You can start from my post above:

`*WAKE=2*TUSB=1*BOOT="!Lbt"!SAVEbt=<u0"On battery"+!XmSr5Xp30fWe1b1d10"Standard video\nCharging 45s..."!9N!9N!9N!9N!9N!S!3N<u0"Shutting down"+!E+!2O<r0"Exiting script"+!1N+!X!R85`

What you need to do:

1 - Take out the part `"Standard video\nCharging 45s..."!9N!9N!9N!9N!9N!S!3N` That'll remove the 45 second delay from power up to recording start.

2 - Change the mode stuff to match the mode you want (see my post above).

I think that's it; it should work then.

As Dave N said, it will create broken files that need to be fixed on the next power up (camera will do this automatically, but make sure you power it up before removing the SD card or downloading the file, or it won't get fixed).

(If you put in a battery then the broken file thing won't be a problem. But you don't have to.)

For more details, see attached.

[Settings commands.pdf](https://github.com/user-attachments/files/15937252/Settings.commands.pdf)

[GoPro Labs command notes for HERO12.txt](https://github.com/user-attachments/files/15937253/GoPro.Labs.command.notes.for.HERO12.txt)

**Reply by dnewman-gpsw** [reply]
Yes all above is correct. For the simplest script for no battery setups, auto capture on power:

```*WAKE=2*TUSB=1*FAST=1*BOOT="!Lbt"!SAVEbt=>u0!S```

After scanning the above, plug in USB power, the camera will boot and immediately start recording. Plug in a battery it will boot and not record automatically. If both cases you can manually stop and start the recording, just a regular camera. Only if power is removed will recording will you need to boot again to repair the file.

**Comment by remltr12** [comment]
*WAKE=2*TUSB=1*BOOT="!Lbt"!SAVEbt=<u0"On battery"+!X_mSr5Xp30fWe1b1d10_<u0"Shutting down"+!E+!2O<r0"Exiting script"+!1N+!X!R85

Change this:
mS	Mode Standard Video
r5X	Mode 5k Extreme 8:7
p30	30 frames/second
fW		Wide lens
e1	Hypersmooth ON
b1	High bitrate (~ 120 Mbps)
d10	10 bits/pixel (HEVC, H.265)

To this:

This is the data to create the QR code when on the Gp pro blockly page - cinematic,2.7K,30 fps,superview,EIS On
Does this make sense? It sort of does to me, but why the generation of duplicate information? And does the order of commands make any difference?

mv2
mv2
r27
r27
p30
fS
fS
fS
e1

WAKE=2TUSB=1*BOOT="!Lbt"!SAVEbt=<u0"On battery"+!X_mSrmv2r27p30fSe1_<u0"Shutting down"+!E+!2O<r0"Exiting script"+!1N+!X!R85

**Reply by nerdfever** [reply]
Yes, makes sense. I don't know why the duplicates are there - you don't need them (they are harmless however).

**Reply by nerdfever** [reply]
Also - no, the order of the *mode commands* doesn't matter. They're just setting modes in the camera. Other things, the order can matter.

I should add - the way that script works, it doesn't change any settings until it boots later on USB power. Don't expect the settings to change when it reads the QR code. (You could write a script that works differently if you wanted.)

**Reply by remltr12** [reply]
OK I created and read the QR code into the GP11. No wake response with or without the battery. Also tried changing On battery to on USB and no wake response.

**Reply by dnewman-gpsw** [reply]
This does work on HERO11, you may have mistyped
![image](https://github.com/gopro/labs/assets/10856084/c9d5f96d-2420-45a0-ad23-0a433693841c)

**Reply by remltr12** [reply]
That works. Thank you very much.

**Reply by remltr12** [reply]
I ran a test and it worked for the most part. I need to be able to set the max time for a video to 5 minutes increments. I assume that mVL is the command I would use,  but how do I tell it 5 minutes? Do I just type mVL5? How about gps and time and is speed possible. All of this on the output video. 
Thanks again.

**Reply by dnewman-gpsw** [reply]
Just set the loop duration in the camera, when using mVL it will remember what you had set.  Not all commands are settable via QR Code., setting the loop duration is one those few missing commands.

**Reply by remltr12** [reply]
Maybe I am misunderstanding the loop mode. I want to continuously record in 5 minute segments instead of one long 10 hour segment for example. When I set loop to 5 minutes it records 5 minutes and then stops recording. What am I missing?

**Reply by dnewman-gpsw** [reply]
That doesn't sound like video loop mode, that seems like duration record.  Two different features. Read up on video loop mode, in the manual or similar.  Labs doesn't modify looping, so it has no documentation on it.

**Comment by dbower44022** [comment]
I have some questions on what happens when a GoPro shuts down due to overheating.   When the camera cools down, will it automatically detect that USB and/or Battery Power has been applied so that scripts will restart? 

If not.... Is there any way to restart a script when a Overheat Shutdown occurs???

**Reply by dnewman-gpsw** [reply]
It is a good question, particularly as it hasn't come up before.  Labs doesn't handle shutdowns due to camera temperature, that is a core camera future, so no wake alarm is set to boot the camera at a later time. So this is no automatic wake event based on time (or temperature.) However if you are using *WAKE=2, and the power is externally cycled the camera will boot and continue operations.

**Reply by dbower44022** [reply]
Is there any capability to monitor temperature in Labs?   If that existed, we could shut down to allow camera to cool rather than and unexpected shutdown.   We are building a construction camera, so it is not easy to cycle the external power as the camera/power must be accessed via ladder.
